File: go/snippets.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

// --- File Operations ---

func ReadFile(path string) (string, error) {
	data, err := os.ReadFile(path)
	return string(data), err
}

func ReadLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func WriteJSON(path string, v interface{}) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0644)
}

// --- String Utils ---

func Slugify(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	s = strings.Join(strings.Fields(s), "-")
	return s
}

func ContainsAny(s string, subs ...string) bool {
	for _, sub := range subs {
		if strings.Contains(s, sub) {
			return true
		}
	}
	return false
}

// --- Functional Utils ---

func Filter[T any](slice []T, fn func(T) bool) []T {
	var result []T
	for _, item := range slice {
		if fn(item) {
			result = append(result, item)
		}
	}
	return result
}

func Map[T, U any](slice []T, fn func(T) U) []U {
	result := make([]U, len(slice))
	for i, item := range slice {
		result[i] = fn(item)
	}
	return result
}

func Unique[T comparable](slice []T) []T {
	seen := make(map[T]bool)
	var result []T
	for _, item := range slice {
		if !seen[item] {
			seen[item] = true
			result = append(result, item)
		}
	}
	return result
}

// --- Timer ---

func TimeIt(name string, fn func()) {
	start := time.Now()
	fn()
	fmt.Printf("%s: %v\n", name, time.Since(start))
}

func main() {
	fmt.Println("Slugify:", Slugify("Hello World Test"))

	nums := []int{1, 2, 2, 3, 3, 4, 5}
	evens := Filter(nums, func(n int) bool { return n%2 == 0 })
	fmt.Println("Evens:", evens)

	doubled := Map(nums, func(n int) int { return n * 2 })
	fmt.Println("Doubled:", doubled)

	fmt.Println("Unique:", Unique([]string{"a", "b", "a", "c"}))

	TimeIt("sleep", func() {
		time.Sleep(100 * time.Millisecond)
	})
}

File: python/snippets.py
"""Useful Python code snippets."""
import os
import sys
import json
from pathlib import Path
from datetime import datetime, timedelta
from typing import Any


# --- File Operations ---

def read_file(path: str, encoding: str = 'utf-8') -> str:
    """Read entire file to string."""
    return Path(path).read_text(encoding=encoding)


def write_file(path: str, content: str, encoding: str = 'utf-8') -> None:
    """Write string to file."""
    Path(path).write_text(content, encoding=encoding)


def read_json(path: str) -> Any:
    """Read JSON file."""
    return json.loads(Path(path).read_text())


def write_json(path: str, data: Any, indent: int = 2) -> None:
    """Write JSON file."""
    Path(path).write_text(json.dumps(data, indent=indent, ensure_ascii=False))


# --- Date/Time ---

def days_ago(n: int) -> datetime:
    """Return datetime N days ago."""
    return datetime.now() - timedelta(days=n)


def format_date(dt: datetime, fmt: str = '%Y-%m-%d %H:%M:%S') -> str:
    """Format datetime to string."""
    return dt.strftime(fmt)


def parse_date(s: str, fmt: str = '%Y-%m-%d') -> datetime:
    """Parse date string."""
    return datetime.strptime(s, fmt)


# --- String Utils ---

def truncate(s: str, length: int = 50, suffix: str = '...') -> str:
    """Truncate string with suffix."""
    if len(s) <= length:
        return s
    return s[:length - len(suffix)] + suffix


def slugify(text: str) -> str:
    """Convert text to URL-friendly slug."""
    import re
    text = text.lower().strip()
    text = re.sub(r'[^\w\s-]', '', text)
    text = re.sub(r'[-\s]+', '-', text)
    return text.strip('-')


def camel_to_snake(name: str) -> str:
    """Convert CamelCase to snake_case."""
    import re
    return re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()


def snake_to_camel(name: str) -> str:
    """Convert snake_case to CamelCase."""
    return ''.join(word.title() for word in name.split('_'))


# --- Environment ---

def get_env(key: str, default: str = '') -> str:
    """Get environment variable with default."""
    return os.environ.get(key, default)


def require_env(key: str) -> str:
    """Get required environment variable or raise."""
    value = os.environ.get(key)
    if value is None:
        raise ValueError(f"Missing required env var: {key}")
    return value


# --- Decorators ---

def retry(max_attempts: int = 3, delay: float = 1.0):
    """Retry decorator with exponential backoff."""
    import time
    def decorator(func):
        def wrapper(*args, **kwargs):
            for attempt in range(max_attempts):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    if attempt == max_attempts - 1:
                        raise
                    time.sleep(delay * (2 ** attempt))
        return wrapper
    return decorator


def timer(func):
    """Time a function execution."""
    import time
    def wrapper(*args, **kwargs):
        start = time.perf_counter()
        result = func(*args, **kwargs)
        elapsed = time.perf_counter() - start
        print(f"{func.__name__}: {elapsed:.4f}s")
        return result
    return wrapper


if __name__ == '__main__':
    print(f"Truncate: {truncate('Hello World this is a long string', 15)}")
    print(f"Slugify: {slugify('Hello World! This is a Test')}")
    print(f"Camel→Snake: {camel_to_snake('MyClassName')}")
    print(f"Snake→Camel: {snake_to_camel('my_class_name')}")
    print(f"Days ago: {days_ago(7).date()}")
